### PR TITLE
[INLONG-6160][Sort] Support dynamic partition for KafkaLoadNode

### LINKED ISSUE #6160: [Feature][Sort] Support dynamic partition for KafkaLoadNode
### Description

Support dynamic partition for KafkaLoadNode when the format of kafka is raw and the 'key.fields' is not specifyed.
This is mainly for some whole database migration scenarios, we assume that the upstream input data is a mixed schema of whole database migration, we ignore the real schema for now, receive the entire record in a binary raw data format, and fetch and parse its schema and data on the kafka sink side, and according to Some data values ​​are dynamically written to related topic partitions.
Dynamic partition writing has some limitations:
1.The upstream data is raw format with a fixed inner format, only support [canal-json|debezium-json] at now
2.The 'key.fields' is not specifyed
3.It needs to specify 'sink.multiple.partition-pattern' and 'sink.multiple.format' for dynamically extracting partition from data

### Use case

_No response_

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by yunqingmoswu**
### Prepare a Pull Request
*(Change the title refer to the following example)*

Title: [INLONG-6160][Sort] Support dynamic partition for KafkaLoadNode

*(The following *XYZ* should be replaced by the actual [GitHub Issue](https://github.com/apache/inlong/issues) number)*

Fixes #6160 

### Motivation

Support dynamic partition for KafkaLoadNode when the format of kafka is raw and the 'key.fields' is not specifyed.
This is mainly for some whole database migration scenarios, we assume that the upstream input data is a mixed schema of whole database migration, we ignore the real schema for now, receive the entire record in a binary raw data format, and fetch and parse its schema and data on the kafka sink side, and according to Some data values ​​are dynamically written to related topic partitions.
Dynamic partition writing has some limitations:
1.The upstream data is raw format with a fixed inner format, only support [canal-json|debezium-json] at now
2.The 'key.fields' is not specifyed
3.It needs to specify 'sink.multiple.partition-pattern' and 'sink.multiple.format' for dynamically extracting partition from data
4.It will extract primary key from raw data as the partition key used hash if the 'sink.multiple.partition-pattern'
 equals 'PRIMARY_KEY' else it will parse partition key from raw data.

### Modifications

1.Add option of 'sink.multiple.partition-pattern'
2.Add partitioner of 'RawDataHashPartitioner'
3.Add method extractValues and extractPrimaryKeyNames and extractPrimaryKeyValues for the mode of 'DynamicSchemaFormat'

### Verifying this change

*(Please pick either of the following options)*

- [ ] This change is a trivial rework/code cleanup without any test coverage.

- [x] This change is already covered by existing tests, such as:
  *(please describe tests)*

- [ ] This change added tests and can be verified as follows:

  *(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a follow-up issue for adding the documentation

### CODE REVIEW

File: inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/load/KafkaLoadNode.java [hunk 1]
```diff
@@ -157,6 +173,12 @@ public Map<String, String> tableOptions() {
                 if (StringUtils.isNotBlank(topicPattern)) {
                     options.put("topic-pattern", topicPattern);
                 }
+                if (StringUtils.isNotBlank(sinkPartitioner)) {
+                    options.put("sink.partitioner", sinkPartitioner);
+                }
+                if (StringUtils.isNotBlank(partitionPattern)) {
+                    options.put("sink.multiple.partition-pattern", partitionPattern);
```
Comment (gong) [inline inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/load/KafkaLoadNode.java:180]: Maybe, optional option  can be passed by properties object. we don't need change construction method.

File: inlong-sort/sort-connectors/kafka/src/main/java/org/apache/inlong/sort/kafka/table/KafkaDynamicTableFactory.java [hunk 2]
```diff
@@ -206,6 +216,18 @@ private static void validateSinkPartitioner(ReadableConfig tableOptions) {
...
+                            || "org.apache.inlong.sort.kafka.partitioner.RawDataHashPartitioner".equals(partitioner))
+                            && (!"raw".equals(tableOptions.getOptional(FORMAT).orElse(null))
+                            || !tableOptions.getOptional(SINK_MULTIPLE_FORMAT).isPresent()
+                            || !tableOptions.getOptional(SINK_MULTIPLE_PARTITION_PATTERN).isPresent()
+                            || tableOptions.getOptional(SINK_MULTIPLE_FORMAT).get().isEmpty()
+                            || tableOptions.getOptional(SINK_MULTIPLE_PARTITION_PATTERN).get().isEmpty())) {
+                        throw new ValidationException(
+                                "Currently 'raw-hash' partitioner only works "
+                                        + "when option 'format' is 'raw' and option 'sink.multiple.format' "
+                                        + "and 'sink.multiple.partition-pattern' is specified.");
```
Comment (gong) [inline inlong-sort/sort-connectors/kafka/src/main/java/org/apache/inlong/sort/kafka/table/KafkaDynamicTableFactory.java:229]: Maybe, add a bool variable.
```java
 boolean notValidRawHashPartitionerConfig = (SINK_PARTITIONER_VALUE_RAW_HASH.equals(partitioner)
                || "org.apache.inlong.sort.kafka.partitioner.RawDataHashPartitioner".equals(partitioner))
                && (!"raw".equals(tableOptions.getOptional(FORMAT).orElse(null))
                || !tableOptions.getOptional(SINK_MULTIPLE_FORMAT).isPresent()
                || !tableOptions.getOptional(SINK_MULTIPLE_PARTITION_PATTERN).isPresent()
                || tableOptions.getOptional(SINK_MULTIPLE_FORMAT).get().isEmpty()
                || tableOptions.getOptional(SINK_MULTIPLE_PARTITION_PATTERN).get().isEmpty());
```

File: inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/load/KafkaLoadNode.java (hunk 1 above)
Comment (yunqingmoswu) [reply inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/load/KafkaLoadNode.java:180]: Mainly to facilitate api integration calls, and for compatibility, a new constructor has been added.

File: inlong-sort/sort-connectors/kafka/src/main/java/org/apache/inlong/sort/kafka/table/KafkaDynamicTableFactory.java (hunk 2 above)
Comment (yunqingmoswu) [reply inlong-sort/sort-connectors/kafka/src/main/java/org/apache/inlong/sort/kafka/table/KafkaDynamicTableFactory.java:229]: The necessity of adding a variable?

Comment (gong) [reply inlong-sort/sort-connectors/kafka/src/main/java/org/apache/inlong/sort/kafka/table/KafkaDynamicTableFactory.java:229]: > 

@yunqingmoswu  Use boolean variables to store results of complicated statements temporarily will increase the code’s readability.[refer to](https://alibaba.github.io/Alibaba-Java-Coding-Guidelines/) 4th item of Flow Control Statements.

**Review by thesumery — APPROVED** [review]

**Review by EMsnap — APPROVED** [review]

File: inlong-sort/sort-connectors/kafka/src/main/java/org/apache/inlong/sort/kafka/table/KafkaDynamicTableFactory.java (hunk 2 above)
Comment (yunqingmoswu) [reply inlong-sort/sort-connectors/kafka/src/main/java/org/apache/inlong/sort/kafka/table/KafkaDynamicTableFactory.java:229]: ok

File: inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/load/KafkaLoadNode.java [hunk 3]
```diff
@@ -112,14 +119,23 @@ public KafkaLoadNode(@JsonProperty("id") String id,
...
+            @Nullable @JsonProperty("partitionPattern") String partitionPattern) {
         super(id, name, fields, fieldRelations, filters, filterStrategy, sinkParallelism, properties);
         this.topic = Preconditions.checkNotNull(topic, "topic is null");
         this.bootstrapServers = Preconditions.checkNotNull(bootstrapServers, "bootstrapServers is null");
         this.format = Preconditions.checkNotNull(format, "format is null");
         this.primaryKey = primaryKey;
         this.sinkMultipleFormat = sinkMultipleFormat;
         this.topicPattern = topicPattern;
+        this.sinkPartitioner = sinkPartitioner;
+        if ("raw-hash".equals(sinkPartitioner)) {
```
[the remark this answers is not in the document]
Comment (yunqingmoswu) [reply inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/load/KafkaLoadNode.java:133]: ok

File: inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/format/CanalJsonDynamicSchemaFormat.java [hunk 4]
```diff
@@ -46,6 +49,18 @@ protected JsonNode getPhysicalData(JsonNode root) {
         return null;
     }
 
+    @Override
+    public List<String> extractPrimaryKeyNames(JsonNode data) {
+        JsonNode pkNamesNode = data.get("pkNames");
```
Comment (healchow) [inline inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/format/CanalJsonDynamicSchemaFormat.java:54]: Suggest extracting a constant String for `pkNames`.

File: inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/format/DebeziumJsonDynamicSchemaFormat.java [hunk 5]
```diff
@@ -46,6 +49,22 @@ protected JsonNode getPhysicalData(JsonNode root) {
         return physicalData;
     }
 
+    @Override
+    public List<String> extractPrimaryKeyNames(JsonNode data) {
+        List<String> pkNames = new ArrayList<>();
+        JsonNode sourceNode = data.get("source");
```
Comment (healchow) [inline inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/format/DebeziumJsonDynamicSchemaFormat.java:55]: Suggest extracting a constant String for `source`.

**Review by healchow — APPROVED** [review]

**Review by gong — APPROVED** [review]

**Pull request merged**